File: impl/go/api_test.go
package jed

// Phase 7: the formal host API (spec/design/api.md) — open/create/commit/close a database file,
// Prepare/queryValues/Exec, the Rows cursor, and the structured-error surface. Files are written
// under t.TempDir(), never the repo tree. Everything runs through the public Database/Session
// surface; the low-level engine is internal.

import (
	"bytes"
	"errors"
	"os"
	"path/filepath"
	"slices"
	"testing"
)

func mustExec(t *testing.T, db dbHandle, sql string) {
	t.Helper()
	if _, err := queryOutcome(db, sql, nil); err != nil {
		t.Fatalf("%q: %v", sql, err)
	}
}

func TestCreateCommitReopenRoundTrips(t *testing.T) {
	path := filepath.Join(t.TempDir(), "round_trip.jed")
	db, err := CreateDatabase(CreateOptions{Path: path, SkipFsync: true})
	if err != nil {
		t.Fatal(err)
	}
	if db.Txid() != 1 { // the initial empty image is committed at create
		t.Fatalf("txid after create = %d want 1", db.Txid())
	}
	mustExec(t, db, "CREATE TABLE t (id i32 PRIMARY KEY, v i32)")
	mustExec(t, db, "INSERT INTO t VALUES (1, 10), (2, 20)") // autocommitted durably
	afterCommit := db.Txid()
	if err := db.Close(); err != nil {
		t.Fatal(err)
	}

	db, err = OpenDatabaseWithOptions(path, OpenOptions{SkipFsync: true})
	if err != nil {
		t.Fatal(err)
	}
	defer db.Close()
	if db.Txid() != afterCommit {
		t.Fatalf("reopened txid = %d want %d", db.Txid(), afterCommit)
	}
	rows := queryRows(t, db, "SELECT id, v FROM t")
	if len(rows) != 2 || rows[0][0].Int != 1 || rows[1][1].Int != 20 {
		t.Fatalf("got %v", rows)
	}
}

func TestOpenMissingFileIs58P01(t *testing.T) {
	path := filepath.Join(t.TempDir(), "nope.jed")
	if _, err := OpenDatabaseWithOptions(path, OpenOptions{SkipFsync: true}); err == nil {
		t.Fatal("expected error")
	} else if ee, ok := err.(*EngineError); !ok || ee.Code() != "58P01" {
		t.Fatalf("code = %v want 58P01", err)
	}
}

func TestCreateOverExistingFileIs58P02(t *testing.T) {
	path := filepath.Join(t.TempDir(), "here.jed")
	db, err := CreateDatabase(CreateOptions{Path: path, SkipFsync: true})
	if err != nil {
		t.Fatal(err)
	}
	if err := db.Close(); err != nil {
		t.Fatal(err)
	}
	if _, err := CreateDatabase(CreateOptions{Path: path, SkipFsync: true}); err == nil {
		t.Fatal("expected error")
	} else if ee, ok := err.(*EngineError); !ok || ee.Code() != "58P02" {
		t.Fatalf("code = %v want 58P02", err)
	}
}

func TestCreateWithCustomPageSizeRoundTrips(t *testing.T) {
	path := filepath.Join(t.TempDir(), "page256.jed")
	db, err := CreateDatabase(CreateOptions{Path: path, PageSize: 256, SkipFsync: true})
	if err != nil {
		t.Fatal(err)
	}
	if db.PageSize() != 256 {
		t.Fatalf("page size = %d want 256", db.PageSize())
	}
	db.Close()
	db, err = OpenDatabaseWithOptions(path, OpenOptions{SkipFsync: true})
	if err != nil {
		t.Fatal(err)
	}
	defer db.Close()
	if db.PageSize() != 256 {
		t.Fatalf("reopened page size = %d want 256", db.PageSize())
	}
}

func TestAutocommitPersistsEachWriteAcrossClose(t *testing.T) {
	// jed autocommits (spec/design/transactions.md §4.1): a write is durable as soon as it
	// succeeds, so it survives a Close with no explicit Commit — the opposite of the original
	// "no autocommit" model this test used to assert.
	path := filepath.Join(t.TempDir(), "autocommit.jed")
	db, err := CreateDatabase(CreateOptions{Path: path, SkipFsync: true})
	if err != nil {
		t.Fatal(err)
	}
	mustExec(t, db, "CREATE TABLE t (id i32 PRIMARY KEY)")
	mustExec(t, db, "INSERT INTO t VALUES (1)") // autocommitted, no explicit commit
	db.Close()

	db, err = OpenDatabaseWithOptions(path, OpenOptions{SkipFsync: true})
	if err != nil {
		t.Fatal(err)
	}
	defer db.Close()
	rows := queryRows(t, db, "SELECT id FROM t")
	if len(rows) != 1 || rows[0][0].Int != 1 {
		t.Fatalf("autocommitted insert must persist, got %v", rows)
	}
}

func TestCommitAndRollbackAreNoopsUnderAutocommit(t *testing.T) {
	// With no explicit transaction open, both are lenient no-op successes (transactions.md §4.2).
	db := memDB().Session(SessionOptions{})
	mustExec(t, db, "CREATE TABLE t (id i32 PRIMARY KEY)")
	mustExec(t, db, "INSERT INTO t VALUES (1)")
	if err := db.Commit(); err != nil {
		t.Fatal(err)
	}
	if err := db.Rollback(); err != nil { // does NOT undo the autocommitted insert
		t.Fatal(err)
	}
	rows := queryRows(t, db, "SELECT id FROM t")
	if len(rows) != 1 || rows[0][0].Int != 1 {
		t.Fatalf("autocommitted row must remain, got %v", rows)
	}
}

func TestPrepareExecuteAndQueryWithParams(t *testing.T) {
	db := memDB().Session(SessionOptions{})
	mustExec(t, db, "CREATE TABLE t (id i32 PRIMARY KEY, v i32)")
	insert, err := db.Prepare("INSERT INTO t VALUES ($1, $2)")
	if err != nil {
		t.Fatal(err)
	}
	if _, err := prepOutcome(insert, []Value{IntValue(1), IntValue(100)}); err != nil {
		t.Fatal(err)
	}
	if _, err := prepOutcome(insert, []Value{IntValue(2), IntValue(200)}); err != nil {
		t.Fatal(err)
	}

	sel, err := db.Prepare("SELECT id, v FROM t WHERE v = $1")
	if err != nil {
		t.Fatal(err)
	}
	rows, err := sel.queryValues([]Value{IntValue(200)})
	if err != nil {
		t.Fatal(err)
	}
	if got := rows.ColumnNames(); len(got) != 2 || got[0] != "id" || got[1] != "v" {
		t.Fatalf("column names = %v", got)
	}
	var collected [][]Value
	for rows.Next() {
		collected = append(collected, rows.Row())
	}
	if len(collected) != 1 || collected[0][0].Int != 2 || collected[0][1].Int != 200 {
		t.Fatalf("got %v", collected)
	}
	if rows.Cost() < 0 {
		t.Fatalf("cost = %d", rows.Cost())
	}
}

// TestQueryOnNonQueryStatementIsTotal locks in the total exec/query seam (spec/design/api.md §11):
// Query on a statement that produces no rows is VALID — it returns a Rows with no columns carrying the
// command tag, and (the effect-then-error bug this fixed) the statement's effect actually lands. A
// write run through Query used to commit and THEN return 42601; now it just succeeds.
func TestQueryOnNonQueryStatementIsTotal(t *testing.T) {
	db := memDB().Session(SessionOptions{})

	// DDL through Query: no columns, no rows, no error, and the table is really created.
	rows, err := db.queryValues("CREATE TABLE t (id i32 PRIMARY KEY, v i32)", nil)
	if err != nil {
		t.Fatalf("Query(CREATE TABLE) errored: %v", err)
	}
	if got := rows.ColumnNames(); len(got) != 0 {
		t.Fatalf("statement Rows has columns %v, want none", got)
	}
	if rows.Next() {
		t.Fatal("statement Rows produced a row")
	}
	if err := rows.Err(); err != nil {
		t.Fatalf("drain error: %v", err)
	}
	_ = rows.Close()

	// A write through Query commits AND carries the affected count on the Rows (no 42601).
	wr, err := db.queryValues("INSERT INTO t VALUES (1, 100), (2, 200)", nil)
	if err != nil {
		t.Fatalf("Query(INSERT) errored: %v", err)
	}
	for wr.Next() {
		t.Fatal("INSERT via Query produced a row")
	}
	if n, ok := wr.RowsAffected(); !ok || n != 2 {
		t.Fatalf("RowsAffected = (%d,%v), want (2,true)", n, ok)
	}
	_ = wr.Close()

	// The insert really landed (proves the write committed, not merely reported).
	sel, err := db.queryValues("SELECT v FROM t ORDER BY id", nil)
	if err != nil {
		t.Fatalf("Query(SELECT) errored: %v", err)
	}
	var got [][]Value
	for sel.Next() {
		got = append(got, append([]Value(nil), sel.Row()...))
	}
	if err := sel.Err(); err != nil {
		t.Fatalf("SELECT drain error: %v", err)
	}
	_ = sel.Close()
	if len(got) != 2 || got[0][0].Int != 100 || got[1][0].Int != 200 {
		t.Fatalf("rows after INSERT via Query = %v", got)
	}
}

func TestErrorsSurfaceWithSQLState(t *testing.T) {
	db := memDB().Session(SessionOptions{})
	if _, err := db.Prepare("SELCT 1"); err == nil {
		t.Fatal("expected error")
	} else if ee, ok := err.(*EngineError); !ok || ee.Code() != "42601" {
		t.Fatalf("code = %v want 42601", err)
	}
}

func TestCommitOnInMemoryIsNoopSuccess(t *testing.T) {
	db := memDB().Session(SessionOptions{})
	mustExec(t, db, "CREATE TABLE t (id i32 PRIMARY KEY)")
	before := db.Txid()
	if err := db.Commit(); err != nil { // no open block -> no-op success, not an error
		t.Fatal(err)
	}
	if db.Txid() != before || db.Path() != "" {
		t.Fatalf("txid=%d (want %d) path=%q want empty", db.Txid(), before, db.Path())
	}
}

func TestTableNamesListsTablesSortedExcludingIndexes(t *testing.T) {
	// The catalog-read surface (api.md §6): canonical names, sorted ascending by
	// lowercased name; secondary indexes are relations but not tables. Uses a single Session so the
	// open BEGIN block is visible across calls (the bare Database conveniences mint a fresh session
	// per call and would not share the block).
	db := memDB().Session(SessionOptions{})
	if got := db.TableNames(); len(got) != 0 {
		t.Fatalf("empty catalog: got %v", got)
	}
	mustCreate(t, db, "CREATE TABLE Zed (id i32 PRIMARY KEY, v i32)")
	mustCreate(t, db, "CREATE TABLE apple (id i32 PRIMARY KEY)")
	mustCreate(t, db, "CREATE INDEX zed_v_idx ON Zed (v)")
	// Sorted by LOWERCASED name (apple < zed), returning the canonical spelling (`Zed`).
	want := []string{"apple", "Zed"}
	if got := db.TableNames(); !slices.Equal(got, want) {
		t.Fatalf("TableNames() = %v, want %v", got, want)
	}
	// The visible snapshot includes an open transaction's working set.
	mustCreate(t, db, "BEGIN")
	mustCreate(t, db, "CREATE TABLE mid (id i32 PRIMARY KEY)")
	if got := db.TableNames(); !slices.Equal(got, []string{"apple", "mid", "Zed"}) {
		t.Fatalf("in-tx TableNames() = %v", got)
	}
	mustCreate(t, db, "ROLLBACK")
	if got := db.TableNames(); !slices.Equal(got, want) {
		t.Fatalf("post-rollback TableNames() = %v, want %v", got, want)
	}
}

func TestRowsAffectedReportsDMLCounts(t *testing.T) {
	// The affected-row count (api.md §4): INSERT/UPDATE/DELETE without RETURNING report
	// how many rows they touched (PostgreSQL's command-tag count); a DML statement that
	// matched nothing reports (0, true); DDL and transaction control report (0, false);
	// DML with RETURNING is a query outcome (its row count is the result's length).
	db := memDB().Session(SessionOptions{})
	affected := func(sql string) (int64, bool) {
		t.Helper()
		out, err := queryOutcome(db, sql, nil)
		if err != nil {
			t.Fatalf("%q: %v", sql, err)
		}
		if out.Kind != outcomeStatement {
			t.Fatalf("%q: expected a statement outcome", sql)
		}
		return out.RowsAffected, out.HasRowsAffected
	}

	if n, ok := affected("CREATE TABLE t (id i32 PRIMARY KEY, v i32)"); ok || n != 0 {
		t.Fatalf("DDL: got (%d, %v) want (0, false)", n, ok)
	}
	if n, ok := affected("INSERT INTO t VALUES (1, 10), (2, 20), (3, 30)"); !ok || n != 3 {
		t.Fatalf("INSERT: got (%d, %v) want (3, true)", n, ok)
	}
	if n, ok := affected("UPDATE t SET v = v + 1 WHERE id <= 2"); !ok || n != 2 {
		t.Fatalf("UPDATE: got (%d, %v) want (2, true)", n, ok)
	}
	if n, ok := affected("DELETE FROM t WHERE id = 3"); !ok || n != 1 {
		t.Fatalf("DELETE: got (%d, %v) want (1, true)", n, ok)
	}
	if n, ok := affected("DELETE FROM t WHERE id = 99"); !ok || n != 0 {
		t.Fatalf("zero-match DELETE: got (%d, %v) want (0, true)", n, ok)
	}
	if n, ok := affected("BEGIN"); ok || n != 0 {
		t.Fatalf("BEGIN: got (%d, %v) want (0, false)", n, ok)
	}
	if n, ok := affected("COMMIT"); ok || n != 0 {
		t.Fatalf("COMMIT: got (%d, %v) want (0, false)", n, ok)
	}

	// INSERT ... SELECT counts the inserted rows; DML with RETURNING is a Query.
	mustExec(t, db, "CREATE TABLE dst (id i32 PRIMARY KEY)")
	if n, ok := affected("INSERT INTO dst SELECT id FROM t"); !ok || n != 2 {
		t.Fatalf("INSERT ... SELECT: got (%d, %v) want (2, true)", n, ok)
	}
	out, err := queryOutcome(db, "DELETE FROM dst RETURNING id", nil)
	if err != nil {
		t.Fatal(err)
	}
	if out.Kind != outcomeQuery || len(out.Rows) != 2 {
		t.Fatalf("RETURNING must yield a query outcome with 2 rows, got kind=%v rows=%d", out.Kind, len(out.Rows))
	}
}

func TestOpenReadOnlyBlocksWritesAndNeverTouchesTheFile(t *testing.T) {
	// Read-only open (api.md §2.1): the handle behaves like PostgreSQL hot standby — every
	// transaction defaults to READ ONLY, an explicit READ WRITE request and any write are
	// 25006, and the file bytes are never touched.
	path := filepath.Join(t.TempDir(), "readonly.jed")
	db, err := CreateDatabase(CreateOptions{Path: path, SkipFsync: true})
	if err != nil {
		t.Fatal(err)
	}
	mustExec(t, db, "CREATE TABLE t (id i32 PRIMARY KEY)")
	mustExec(t, db, "INSERT INTO t VALUES (1)")
	if err := db.Close(); err != nil {
		t.Fatal(err)
	}
	before, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}

	rodb, err := OpenDatabaseWithOptions(path, OpenOptions{ReadOnly: true, SkipFsync: true})
	if err != nil {
		t.Fatal(err)
	}
	if !rodb.ReadOnly() {
		t.Fatal("handle must report read-only")
	}
	// A single stateful session so an open block spans calls (a read-only block, then poisoning).
	s := rodb.Session(SessionOptions{})
	wantCode := func(sql, code string) {
		t.Helper()
		_, err := queryOutcome(s, sql, nil)
		var ee *EngineError
		if !errors.As(err, &ee) || ee.Code() != code {
			t.Fatalf("%q: got %v, want %s", sql, err, code)
		}
	}

	// Reads work — bare and inside an explicit block (plain BEGIN defaults to READ ONLY here).
	out, err := queryOutcome(s, "SELECT id FROM t", nil)
	if err != nil || len(out.Rows) != 1 {
		t.Fatalf("read on a read-only handle: %v", err)
	}
	mustExec(t, s, "BEGIN")
	mustExec(t, s, "SELECT id FROM t")
	mustExec(t, s, "COMMIT")

	// Autocommit writes are 25006 (the implicit transaction is read-only)...
	wantCode("INSERT INTO t VALUES (2)", "25006")
	// ...as are writes inside a block (which then poisons, like any in-block error)...
	mustExec(t, s, "BEGIN")
	wantCode("DELETE FROM t", "25006")
	wantCode("SELECT id FROM t", "25P02")
	mustExec(t, s, "ROLLBACK")
	// ...and an explicit READ WRITE request, via SQL or the host API.
	wantCode("BEGIN READ WRITE", "25006")
	var ee *EngineError
	if err := s.Begin(true); !errors.As(err, &ee) || ee.Code() != "25006" {
		t.Fatalf("Begin(true) on a read-only handle: %v", err)
	}
	if err := rodb.View(func(tx *Transaction) error { _, err := tx.queryValues("SELECT id FROM t", nil); return err }); err != nil {
		t.Fatalf("View on a read-only handle: %v", err)
	}
	err = rodb.Update(func(tx *Transaction) error { _, err := queryOutcome(tx, "DELETE FROM t", nil); return err })
	if !errors.As(err, &ee) || ee.Code() != "25006" {
		t.Fatalf("Update on a read-only handle: %v", err)
	}
	s.Close()
	if err := rodb.Close(); err != nil {
		t.Fatal(err)
	}

	// The file is byte-identical after the whole read-only session.
	after, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(before, after) {
		t.Fatal("a read-only session must not change the file")
	}

	// A normal reopen is writable again.
	wdb, err := OpenDatabaseWithOptions(path, OpenOptions{SkipFsync: true})
	if err != nil {
		t.Fatal(err)
	}
	defer wdb.Close()
	if wdb.ReadOnly() {
		t.Fatal("a normal open must not be read-only")
	}
	mustExec(t, wdb, "INSERT INTO t VALUES (2)")
}

File: impl/ts/tests/api.test.ts
// Phase 7: the formal host API (spec/design/api.md) — open/create/commit/close a database file,
// prepare/execute/query, the Rows cursor, and the structured-error surface. Files are written
// under a fresh mkdtemp dir, never the repo tree.

import assert from "node:assert/strict";
import { mkdtempSync, readFileSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { test } from "node:test";
import {
  begin,
  close,
  commit,
  create,
  Engine,
  EngineError,
  execute,
  executePrepared,
  intValue,
  open,
  prepare,
  query,
  queryPrepared,
  rollback,
  update,
  view,
} from "../src/tooling.ts";
import type { Value } from "../src/lib.ts";

function tmpDir(): string {
  return mkdtempSync(join(tmpdir(), "jed-"));
}

test("create → commit → reopen round-trips", () => {
  const dir = tmpDir();
  try {
    const path = join(dir, "round_trip.jed");
    const db = create(path);
    assert.equal(db.txid, 1n); // the initial empty image is committed at create
    execute(db, "CREATE TABLE t (id i32 PRIMARY KEY, v i32)");
    execute(db, "INSERT INTO t VALUES (1, 10), (2, 20)");
    commit(db);
    const afterCommit = db.txid;
    close(db);

    const db2 = open(path);
    assert.equal(db2.txid, afterCommit);
    const o = execute(db2, "SELECT id, v FROM t");
    assert.equal(o.kind, "query");
    if (o.kind === "query") assert.equal(o.rows.length, 2);
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});

test("open missing file is 58P01", () => {
  const dir = tmpDir();
  try {
    let code = "";
    try {
      open(join(dir, "nope.jed"));
    } catch (e) {
      if (e instanceof EngineError) code = e.code();
    }
    assert.equal(code, "58P01");
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});

test("create over existing file is 58P02", () => {
  const dir = tmpDir();
  try {
    const path = join(dir, "here.jed");
    create(path);
    let code = "";
    try {
      create(path);
    } catch (e) {
      if (e instanceof EngineError) code = e.code();
    }
    assert.equal(code, "58P02");
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});

test("create with custom page size round-trips", () => {
  const dir = tmpDir();
  try {
    const path = join(dir, "page256.jed");
    const db = create(path, { pageSize: 256 });
    assert.equal(db.pageSize, 256);
    close(db);
    const bytes = readFileSync(path);
    const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    assert.equal(dv.getUint32(8, false), 256);
    assert.equal(open(path).pageSize, 256);
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});

test("autocommit persists each write across close", () => {
  // jed autocommits (spec/design/transactions.md §4.1): a write is durable as soon as it
  // succeeds, so it survives a close with no explicit commit — the opposite of the original
  // "no autocommit" model this test used to assert.
  const dir = tmpDir();
  try {
    const path = join(dir, "autocommit.jed");
    const db = create(path);
    execute(db, "CREATE TABLE t (id i32 PRIMARY KEY)");
    execute(db, "INSERT INTO t VALUES (1)"); // autocommitted, no explicit commit
    close(db);

    const db2 = open(path);
    const o = execute(db2, "SELECT id FROM t");
    assert.equal(o.kind, "query");
    if (o.kind === "query") {
      assert.equal(o.rows.length, 1);
      assert.deepEqual(o.rows[0][0], intValue(1n));
    }
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});

test("commit and rollback are no-ops under autocommit", () => {
  // With no explicit transaction open, both are lenient no-op successes (transactions.md §4.2).
  const db = new Engine();
  execute(db, "CREATE TABLE t (id i32 PRIMARY KEY)");
  execute(db, "INSERT INTO t VALUES (1)");
  commit(db);
  rollback(db); // does NOT undo the autocommitted insert
  const o = execute(db, "SELECT id FROM t");
  assert.equal(o.kind, "query");
  if (o.kind === "query") {
    assert.equal(o.rows.length, 1);
    assert.deepEqual(o.rows[0][0], intValue(1n));
  }
});

test("prepare → execute → query with params, iterating rows", () => {
  const db = new Engine();
  execute(db, "CREATE TABLE t (id i32 PRIMARY KEY, v i32)");
  const insert = prepare(db, "INSERT INTO t VALUES ($1, $2)");
  executePrepared(db, insert, [intValue(1n), intValue(100n)]);
  executePrepared(db, insert, [intValue(2n), intValue(200n)]);

  const sel = prepare(db, "SELECT id, v FROM t WHERE v = $1");
  const r = queryPrepared(db, sel, [intValue(200n)]);
  assert.deepStrictEqual(r.columnNames, ["id", "v"]);
  const collected: Value[][] = [];
  for (const row of r) collected.push(row);
  assert.equal(collected.length, 1);
  assert.equal(collected[0]![0]!.kind === "int" ? collected[0]![0]!.int : -1n, 2n);
  assert.ok(r.cost >= 0n);
});

test("one-shot query iterates rows", () => {
  const db = new Engine();
  execute(db, "CREATE TABLE t (id i32 PRIMARY KEY)");
  execute(db, "INSERT INTO t VALUES (1), (2), (3)");
  const ids: bigint[] = [];
  for (const row of query(db, "SELECT id FROM t")) {
    if (row[0]!.kind === "int") ids.push(row[0]!.int);
  }
  assert.deepStrictEqual(ids, [1n, 2n, 3n]);
});

test("query on a non-query statement is total (no rows, statement still runs)", () => {
  // `query` is the one total seam (spec/design/api.md §11): a non-query statement is observably a Rows
  // with no output columns — NOT a "use execute" throw (the removed effect-then-error surprise). The
  // DDL still takes effect, and a write exposes its command tag via rowsAffected.
  const db = new Engine();
  const ddl = query(db, "CREATE TABLE t (id i32 PRIMARY KEY)");
  assert.equal(ddl.columnNames.length, 0);
  assert.deepStrictEqual([...ddl], []);
  assert.equal(ddl.rowsAffected, null); // DDL carries no row count
  ddl.close();
  const ins = query(db, "INSERT INTO t VALUES (1), (2)");
  assert.deepStrictEqual([...ins], []);
  assert.equal(ins.rowsAffected, 2);
  ins.close();
});

test("errors surface with SQLSTATE", () => {
  const db = new Engine();
  let code = "";
  try {
    prepare(db, "SELCT 1");
  } catch (e) {
    if (e instanceof EngineError) code = e.code();
  }
  assert.equal(code, "42601");
});

test("commit on in-memory db is a no-op success", () => {
  const db = new Engine();
  execute(db, "CREATE TABLE t (id i32 PRIMARY KEY)");
  commit(db); // no path -> no-op, not an error
  assert.equal(db.txid, 0n);
  assert.equal(db.path, null);
});

test("rowsAffected reports DML counts", () => {
  // The affected-row count (api.md §4): INSERT/UPDATE/DELETE without RETURNING report
  // how many rows they touched (PostgreSQL's command-tag count); a DML statement that
  // matched nothing reports 0; DDL and transaction control report null; DML with
  // RETURNING is a query outcome (its row count is the result's length).
  const db = new Engine();
  const affected = (sql: string): number | null => {
    const out = execute(db, sql);
    assert.equal(out.kind, "statement", sql);
    return out.kind === "statement" ? out.rowsAffected : null;
  };

  assert.equal(affected("CREATE TABLE t (id i32 PRIMARY KEY, v i32)"), null);
  assert.equal(affected("INSERT INTO t VALUES (1, 10), (2, 20), (3, 30)"), 3);
  assert.equal(affected("UPDATE t SET v = v + 1 WHERE id <= 2"), 2);
  assert.equal(affected("DELETE FROM t WHERE id = 3"), 1);
  assert.equal(affected("DELETE FROM t WHERE id = 99"), 0);
  assert.equal(affected("BEGIN"), null);
  assert.equal(affected("COMMIT"), null);

  // INSERT ... SELECT counts the inserted rows; DML with RETURNING is a Query.
  execute(db, "CREATE TABLE dst (id i32 PRIMARY KEY)");
  assert.equal(affected("INSERT INTO dst SELECT id FROM t"), 2);
  const out = execute(db, "DELETE FROM dst RETURNING id");
  assert.equal(out.kind, "query");
  if (out.kind === "query") assert.equal(out.rows.length, 2);
});

test("open read-only blocks writes and never touches the file", () => {
  // Read-only open (api.md §2.1): the handle behaves like PostgreSQL hot standby — every
  // transaction defaults to READ ONLY, an explicit READ WRITE request and any write are
  // 25006, and the file bytes are never touched.
  const dir = tmpDir();
  try {
    const path = join(dir, "readonly.jed");
    let db = create(path);
    execute(db, "CREATE TABLE t (id i32 PRIMARY KEY)");
    execute(db, "INSERT INTO t VALUES (1)");
    close(db);
    const before = readFileSync(path);

    db = open(path, { readOnly: true });
    assert.equal(db.readOnly, true);
    const wantCode = (fn: () => void, code: string) => {
      try {
        fn();
      } catch (e) {
        if (e instanceof EngineError) {
          assert.equal(e.code(), code);
          return;
        }
        throw e;
      }
      assert.fail(`expected ${code}`);
    };

    // Reads work — bare and inside an explicit block (plain BEGIN defaults to READ ONLY here).
    const out = execute(db, "SELECT id FROM t");
    assert.equal(out.kind === "query" && out.rows.length, 1);
    execute(db, "BEGIN");
    execute(db, "SELECT id FROM t");
    execute(db, "COMMIT");

    // Autocommit writes are 25006 (the implicit transaction is read-only)...
    wantCode(() => execute(db, "INSERT INTO t VALUES (2)"), "25006");
    // ...as are writes inside a block (which then poisons, like any in-block error)...
    execute(db, "BEGIN");
    wantCode(() => execute(db, "DELETE FROM t"), "25006");
    wantCode(() => execute(db, "SELECT id FROM t"), "25P02");
    execute(db, "ROLLBACK");
    // ...and an explicit READ WRITE request, via SQL or the host API.
    wantCode(() => execute(db, "BEGIN READ WRITE"), "25006");
    wantCode(() => begin(db, true), "25006");
    view(db, (tx) => tx.query("SELECT id FROM t"));
    wantCode(() => update(db, (tx) => tx.execute("DELETE FROM t")), "25006");
    close(db);

    // The file is byte-identical after the whole read-only session.
    assert.deepStrictEqual(readFileSync(path), before);

    // A normal reopen is writable again.
    db = open(path);
    assert.equal(db.readOnly, false);
    execute(db, "INSERT INTO t VALUES (2)");
    close(db);
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});
